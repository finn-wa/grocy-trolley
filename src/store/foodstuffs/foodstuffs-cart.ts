import { deleteForJson, getForJson, postForJson } from "@grocy-trolley/utils/fetch-utils";
import {
  FoodstuffsAuthService,
  FoodstuffsBaseProduct,
  FoodstuffsCartProduct,
  FoodstuffsStore,
  ListProductRef,
  ProductsSnapshot,
} from ".";
import { FoodstuffsRestService } from "./foodstuffs-rest-service";

export class FoodstuffsCartService extends FoodstuffsRestService {
  constructor(authService: FoodstuffsAuthService) {
    super(authService);
  }

  getCart(): Promise<FoodstuffsCart> {
    return getForJson(this.buildUrl("Cart/Index"), this.authHeaders().acceptJson().build());
  }

  async clearCart(): Promise<{ success: true }> {
    const response: { success: boolean } = await deleteForJson(
      this.buildUrl("Cart/Clear"),
      this.authHeaders().acceptJson().build()
    );
    if (!response.success) {
      throw new Error(`Failed to clear cart: ${response}`);
    }
    return response as { success: true };
  }

  addProductsToCart(products: CartProductRef[]): Promise<FoodstuffsCart> {
    return postForJson(
      this.buildUrl("Cart/Index"),
      this.authHeaders().contentTypeJson().acceptJson().build(),
      { products }
    );
  }
}

export interface FoodstuffsCart {
  products: FoodstuffsCartProduct[];
  unavailableProducts: FoodstuffsCartProduct[];
  subtotal: number;
  promoCodeDiscount: number;
  saving: number;
  serviceFee: number;
  bagFee: number;
  store: FoodstuffsStore;
  orderNumber: number;
  allowSubstitutions: boolean;
  wasRepriced: boolean;
}

export function getCartTitle(cart: FoodstuffsCart): string {
  const now = new Date();
  const date = `${now.getFullYear()}-${now.getMonth() + 1}-${now.getDate()}`;
  const numItems = cart.products.length + cart.unavailableProducts.length;
  return `Cart | ${date} | ${numItems} item${numItems === 1 ? "" : "s"}`;
}

export interface CartProductRef extends ListProductRef {
  restricted: boolean;
}

export function toCartProductRef(product: FoodstuffsBaseProduct): CartProductRef {
  return {
    productId: product.productId,
    quantity: product.quantity,
    restricted: product.restricted,
    saleType: product.sale_type,
  };
}

export function snapshotToCartProductRefs(products: ProductsSnapshot) {
  return [...products.unavailableProducts, ...products.products].map((product) =>
    toCartProductRef(product)
  );
}
