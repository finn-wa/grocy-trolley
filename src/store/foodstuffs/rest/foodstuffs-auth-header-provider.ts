import { AppTokens } from "@gt/app/di";
import { AuthHeaderProvider } from "@gt/store/shared/rest/auth-header-provider";
import { LoginDetails } from "@gt/store/shared/rest/login-details.model";
import { CacheServiceFactory } from "@gt/utils/cache";
import { HeadersBuilder } from "@gt/utils/headers";
import { Logger, prettyPrint } from "@gt/utils/logger";
import { getHeadersFromRequest } from "@gt/utils/playwright";
import { RestService } from "@gt/utils/rest";
import { Browser, Page } from "playwright";
import { inject, Lifecycle, scoped } from "tsyringe";
import { FoodstuffsTokens } from "../foodstuffs-di";
import { List } from "../lists/foodstuffs-list.model";
import { PAKNSAVE_URL } from "../models";

class FoodstuffsAuthHeaderTestService extends RestService {
  protected readonly baseUrl = `${PAKNSAVE_URL}/CommonApi`;
  protected readonly logger = new Logger(this.constructor.name);

  constructor(private readonly headers: Headers) {
    super();
  }

  getCart() {
    return this.getAndParse(this.buildUrl("/Cart/Index"), {
      headers: new HeadersBuilder(this.headers).acceptJson().build(),
    });
  }

  async createList(name: string): Promise<List> {
    return this.putAndParse(this.buildUrl("ShoppingLists/CreateList", { name }), {
      headers: new HeadersBuilder(this.headers).acceptJson().build(),
    });
  }

  async deleteList(id: string | number): Promise<Response> {
    return this.delete(this.buildUrl(`ShoppingLists/DeleteList/${id}`), {
      headers: new HeadersBuilder(this.headers).acceptJson().build(),
    });
  }
}

/**
 * Uses Playwright to perform Foodstuffs requests from a browser. Necessary
 * because Cloudflare now blocks requests that are not sent from a browser.
 */
@scoped(Lifecycle.ContainerScoped)
export class FoodstuffsAuthHeaderProvider extends AuthHeaderProvider {
  protected readonly logger = new Logger(this.constructor.name);
  protected readonly headersFilter = {
    allowed: [
      "host",
      "user-agent",
      "accept-language",
      "accept-encoding",
      "referer",
      "dnt",
      "cookie",
      "sec-fetch-dest",
      "sec-fetch-mode",
      "sec-fetch-site",
    ],
    disallowed: [
      "accept",
      "content-type",
      "x-newrelic-id",
      "newrelic",
      "traceparent",
      "tracestate",
      "__requestverificationtoken",
      "connection",
      "pragma",
      "cache-control",
      "te",
    ],
  };

  constructor(
    @inject(AppTokens.cacheServiceFactory)
    cacheServiceFactory: CacheServiceFactory<{ headers: Record<string, string[]> }>,
    @inject("BrowserLoader") browserLoader: () => Promise<Browser>,
    @inject(FoodstuffsTokens.loginDetails) loginDetails?: LoginDetails
  ) {
    super(cacheServiceFactory, "foodstuffs", browserLoader, loginDetails);
  }

  async login(page: Page): Promise<Headers> {
    await page.goto(`${PAKNSAVE_URL}/shop`);
    let cartRequest = await page.waitForRequest(`${PAKNSAVE_URL}/CommonApi/Cart/Index`);

    if (this.loginDetails && !(await this.isLoggedIn(page))) {
      this.logger.info("Logging in to foodstuffs...");
      await page.click('button[id="login-form"]');
      await page.fill('input[id="login-email"]', this.loginDetails.email);
      await page.fill('input[id="login-password"]', this.loginDetails.password);
      await page.click("button.login-form-submit");
      cartRequest = await page.waitForRequest(`${PAKNSAVE_URL}/CommonApi/Cart/Index`);
      // page refreshes after submission
      await page.waitForLoadState("networkidle");
      // search box seems to pop in last
      await page.locator('input[aria-label="Search products"]').waitFor();
    }
    return getHeadersFromRequest(cartRequest);
  }

  private async isLoggedIn(page: Page) {
    type DataLayer = { loginState?: "loggedIn" | "guest" }[];
    const loginState = await page.evaluate(() => {
      const dataLayer = (globalThis as unknown as { dataLayer: DataLayer }).dataLayer;
      return dataLayer.find((entry) => "loginState" in entry)?.loginState;
    });
    return loginState === "loggedIn";
  }

  protected async isValid(headers: Headers): Promise<boolean> {
    const testService = new FoodstuffsAuthHeaderTestService(headers);
    try {
      const cart = await testService.getCart();
      if (!cart || typeof cart !== "object" || Object.keys(cart).length === 0) {
        // We expect getCart to throw an error if it failed, but if it returns ok check the body
        throw new Error("Test getCart request failed: " + prettyPrint(cart));
      }
      const list = await testService.createList("[temp]");
      if (
        !list ||
        typeof list !== "object" ||
        Object.keys(list).length === 0 ||
        !("listId" in list)
      ) {
        throw new Error("Test createList request failed: " + prettyPrint(cart));
      }
      await testService.deleteList(list.listId);
    } catch (error) {
      this.logger.debug("Headers failed validation", headers, error);
      return false;
    }
    return true;
  }
}

export interface FoodstuffsUserProfile {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  agreedToMarketing: boolean;
  agreedToTermsAndConditions: boolean;
  clubCardUser: boolean;
}

export interface LoginResponse {
  success: boolean;
  userProfile: FoodstuffsUserProfile;
}
