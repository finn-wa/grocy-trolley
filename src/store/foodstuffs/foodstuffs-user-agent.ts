import { getEnv } from "env";
import { access } from "fs/promises";
import { Headers, Response as NodeResponse } from "node-fetch";
import { Browser, BrowserContext, JSHandle, Page } from "playwright";
import { APPLICATION_JSON, headers } from "utils/headers-builder";
import { Logger, prettyPrint } from "utils/logger";
import { PAKNSAVE_URL } from "./foodstuffs.model";

export class FoodstuffsUserAgent {
  protected readonly logger = new Logger(this.constructor.name);
  private readonly loginDetails: LoginDetails;
  private context?: BrowserContext;
  private page?: Page;

  // TODO: support creating a user agent without logging in. for unauthenticated search requests
  constructor(
    private readonly browser: Browser,
    loginDetails?: LoginDetails,
    private readonly storageStatePath = "src/resources/cache/playwright.json"
  ) {
    if (loginDetails) {
      this.loginDetails = loginDetails;
    } else {
      const { PAKNSAVE_EMAIL: email, PAKNSAVE_PASSWORD: password } = getEnv();
      this.loginDetails = { email, password };
    }
  }

  /**
   * Performs a fetch request as a logged in user on the PAK'n'SAVE website.
   * @param method Request method
   * @param url Request URL
   * @param headers Request headers
   * @param body Request body
   * @returns Request response
   */
  async fetchAsUser(
    method: string,
    url: string,
    headers?: Headers,
    body?: any
  ): Promise<NodeResponse> {
    const page = await this.getPage();
    return this.fetchWithPage(page, method, url, headers, body);
  }

  /**
   * Performs a fetch request from the browser using Playwright. Necessary
   * because Cloudflare now blocks requests that are not sent from a browser.
   * WARNING: not every method on the returned Response is callable, see
   * {@link FoodstuffsRequest}.
   * @param page Playwright page instance to use to perform the request
   * @param method Request method
   * @param url Request URL
   * @param headers Request headers
   * @param body Request body
   * @returns Request response
   */
  private async fetchWithPage(
    page: Page,
    method: string,
    url: string,
    headers?: Headers,
    body?: BodyInit | any
  ): Promise<NodeResponse> {
    this.logger.debug(`${method} ${url}`);
    if (headers) {
      this.logger.trace(headers);
    }
    if (body) {
      this.logger.trace(body);
    }
    const contentType = headers?.get("content-type");
    if (contentType === APPLICATION_JSON && body) {
      body = JSON.stringify(body);
    }
    const responseHandle = await page.evaluateHandle(
      async ({ url, method, headers, body }) =>
        fetch(url, {
          credentials: "include",
          referrer: "https://www.paknsave.co.nz/shop",
          mode: "cors",
          method,
          headers,
          body,
        }),
      { url, method, headers, body }
    );
    // Response is not serialisable, so we must serialise it
    const response = await responseHandle.evaluate((response, props) => {
      const responseJson: Record<string, any> = Object.fromEntries(
        props.map((key) => [key, response[key]])
      );
      responseJson.headers = Object.fromEntries(responseJson.headers.entries());
      return responseJson as SerialisedResponse;
    }, SERIALISABLE_RESPONSE_PROPS);
    if (!response.ok) {
      throw new Error(`Response not OK: ${prettyPrint(response)}`);
    }
    this.logger.trace("Response: " + prettyPrint(response));
    return new FoodstuffsResponse(responseHandle, response);
  }

  /**
   * Returns the login page. Creates a new page and logs in if needed.
   * @returns Login page
   */
  private async getPage(): Promise<Page> {
    if (this.page) {
      return this.page;
    }
    const context = await this.getContext();
    const page = await context.newPage();
    await page.goto(PAKNSAVE_URL + "shop");
    // Check if already logged in
    try {
      await this.fetchWithPage(
        page,
        "GET",
        `${PAKNSAVE_URL}/CommonApi/Account/GetUserProfile`,
        headers().acceptJson().build()
      );
    } catch (error) {
      await page.click('button[id="login-form"]');
      await page.fill('input[id="login-email"]', this.loginDetails.email);
      await page.fill('input[id="login-password"]', this.loginDetails.password);
      await page.click("button.login-form-submit");
      // page refreshes after submission
      await page.waitForLoadState("networkidle");
      // search box seems to pop in last
      await page.locator('input[aria-label="Search products"]').waitFor();
      // save logged in state
      await page.context().storageState({ path: this.storageStatePath });
    }
    this.page = page;
    return this.page;
  }

  /**
   * Returns the browser context with cached storage state (if found).
   * Creates a new browser context if needed.
   * @returns Browser context
   */
  private async getContext(): Promise<BrowserContext> {
    if (!this.context) {
      try {
        await access(this.storageStatePath);
        // No error means that cached storageState exists
        this.context = await this.browser.newContext({ storageState: this.storageStatePath });
      } catch (error) {
        this.logger.info("No storageState found, creating new context");
        this.context = await this.browser.newContext();
      }
    }
    return this.context;
  }
}

export interface LoginDetails {
  email: string;
  password: string;
}

export interface FoodstuffsUserProfile {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  agreedToMarketing: boolean;
  agreedToTermsAndConditions: boolean;
  clubCardUser: boolean;
}

export interface LoginResponse {
  success: boolean;
  userProfile: FoodstuffsUserProfile;
}

/**
 * Response backed by a Playwright JSHandle browser fetch Response. json()
 * and text() extraction methods are available.
 */
class FoodstuffsResponse implements NodeResponse {
  readonly bodyUsed: boolean;
  readonly headers: Headers;
  readonly ok: boolean;
  readonly redirected: boolean;
  readonly status: number;
  readonly statusText: string;
  readonly type: ResponseType;
  readonly url: string;

  constructor(
    private readonly responseHandle: JSHandle<Response>,
    responseJson: SerialisedResponse
  ) {
    this.bodyUsed = responseJson.bodyUsed;
    this.ok = responseJson.ok;
    this.redirected = responseJson.redirected;
    this.status = responseJson.status;
    this.statusText = responseJson.statusText;
    this.type = responseJson.type;
    this.url = responseJson.url;
    this.headers = new Headers(responseJson.headers);
  }

  json(): Promise<any> {
    return this.responseHandle.evaluate((response) => response.json());
  }

  text(): Promise<string> {
    return this.responseHandle.evaluate((response) => response.text());
  }

  get body(): NodeJS.ReadableStream | null {
    throw new Error("Cannot access body of FoodstuffsResponse directly");
  }
  get size(): number {
    throw new Error("Cannot access size of FoodstuffsResponse directly");
  }
  buffer(): Promise<Buffer> {
    throw new Error("Method not implemented.");
  }
  clone(): NodeResponse {
    throw new Error("Method not implemented.");
  }
  arrayBuffer(): Promise<ArrayBuffer> {
    throw new Error("Method not implemented.");
  }
  blob(): Promise<Blob> {
    throw new Error("Method not implemented.");
  }
  formData(): Promise<FormData> {
    throw new Error("Method not implemented.");
  }
}

const UNSERIALISABLE_RESPONSE_PROPS = [
  "arrayBuffer",
  "blob",
  "formData",
  "json",
  "text",
  "clone",
  "body",
] as const;

const SERIALISABLE_RESPONSE_PROPS = [
  "headers",
  "ok",
  "redirected",
  "status",
  "statusText",
  "type",
  "url",
] as const;

interface SerialisedResponse
  extends Omit<Response, typeof UNSERIALISABLE_RESPONSE_PROPS[number] | "headers"> {
  headers: Record<string, string>;
}
