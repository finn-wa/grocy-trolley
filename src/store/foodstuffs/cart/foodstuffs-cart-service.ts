import { Logger, prettyPrint } from "@gt/utils/logger";
import { RequestError } from "@gt/utils/rest";
import prompts from "prompts";
import { FoodstuffsCartController } from "./foodstuffs-cart-controller";
import { CartProductRef, FoodstuffsCart } from "./foodstuffs-cart.model";

export class FoodstuffsCartService {
  protected readonly logger = new Logger(this.constructor.name);

  constructor(private readonly controller: FoodstuffsCartController) {}

  getCart = () => this.controller.getCart();

  clearCart = () => this.controller.clearCart();

  async addProductsToCart(products: CartProductRef[]): Promise<FoodstuffsCart> {
    try {
      const cart = await this.controller.postProducts(products);
      return cart;
    } catch (error) {
      this.logger.error("Failed to add products to cart. Falling back to chunks.");
      if (error instanceof RequestError) {
        this.logger.debug(await error.response.text());
      }
    }
    const iter = products[Symbol.iterator]();
    let chunk: CartProductRef[];
    do {
      chunk = Array.from(
        { length: 5 },
        () => iter.next().value as CartProductRef | undefined
      ).filter((p): p is CartProductRef => !!p);
      try {
        await this.controller.postProducts(chunk);
      } catch (error) {
        await this.addProductsToCartIndividually(chunk);
      }
    } while (chunk.length === 5);
    return this.getCart();
  }

  private async addProductsToCartIndividually(products: CartProductRef[]): Promise<FoodstuffsCart> {
    for (const product of products) {
      this.logger.debug("Adding product " + product.productId);
      try {
        await this.controller.postProducts([product]);
      } catch (error) {
        this.logger.error("Failed to add product to cart!\n" + prettyPrint(product));
        if (error instanceof RequestError) {
          this.logger.debug(await error.response.text());
        }
        const response = await prompts([
          { name: "resume", type: "confirm", message: "Resume adding products?" },
        ]);
        if (!response.resume) {
          throw error;
        }
      }
    }
    return this.getCart();
  }
}
