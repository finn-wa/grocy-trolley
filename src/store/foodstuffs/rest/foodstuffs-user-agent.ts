import { LoginDetails } from "@gt/store/shared/rest/login-details.model";
import { StoreUserAgent } from "@gt/store/shared/rest/store-user-agent";
import { APPLICATION_JSON, headersToRaw } from "@gt/utils/headers";
import { Logger, prettyPrint } from "@gt/utils/logger";
import { BrowserContext, JSHandle, Page } from "playwright-firefox";
import { PAKNSAVE_URL } from "../models";

/**
 * Uses Playwright to perform Foodstuffs requests from a browser. Necessary
 * because Cloudflare now blocks requests that are not sent from a browser.
 */
export class FoodstuffsUserAgent extends StoreUserAgent {
  public readonly storeName = "foodstuffs";
  protected readonly logger = new Logger(this.constructor.name);

  async init(context: BrowserContext): Promise<{ page: Page; headers: Headers }> {
    const page = await context.newPage();
    await page.goto(`${PAKNSAVE_URL}/shop`);
    let cartRequest = await page.waitForRequest(`${PAKNSAVE_URL}/CommonApi/Cart/Index`);

    if (this.loginDetails && !(await this.isLoggedIn(page))) {
      this.logger.info("Logging in to foodstuffs...");
      await page.click('button[id="login-form"]');
      await page.fill('input[id="login-email"]', this.loginDetails.email);
      await page.fill('input[id="login-password"]', this.loginDetails.password);
      await page.click("button.login-form-submit");
      cartRequest = await page.waitForRequest(`${PAKNSAVE_URL}/CommonApi/Cart/Index`);
      // page refreshes after submission
      await page.waitForLoadState("networkidle");
      // search box seems to pop in last
      await page.locator('input[aria-label="Search products"]').waitFor();
    }
    const headers = await this.getHeadersFromRequest(cartRequest);
    return { page, headers };
  }

  /**
   * Creates a new agent with the same browser but new login details
   * @param loginDetails Login details for the new agent, or null to omit them
   * @returns the new agent
   */
  clone(loginDetails: LoginDetails | null): FoodstuffsUserAgent {
    return new FoodstuffsUserAgent(this.browserLoader, loginDetails ?? undefined);
  }

  /**
   * Performs a fetch request as a logged in user on the PAK'n'SAVE website.
   * @param method Request method
   * @param url Request URL
   * @param headers Request headers
   * @param body Request body
   * @returns Request response
   */
  async fetchWithBrowser(
    method: string,
    url: string,
    headers?: Headers,
    body?: BodyInit
  ): Promise<Response> {
    const page = await this.getLoginPage();
    const contentType = headers?.get("content-type");
    if (contentType === APPLICATION_JSON && body) {
      body = JSON.stringify(body);
    }
    const rawHeaders = headers ? headersToRaw(headers) : undefined;
    const responseHandle = await page.evaluateHandle(
      async ({ url, method, rawHeaders, body }) =>
        fetch(url, {
          credentials: this.loginDetails ? "include" : undefined,
          referrer: "https://www.paknsave.co.nz/shop",
          mode: "cors",
          method,
          // Record<string, string[]> is not in HeadersInit but it works in firefox at least
          headers: rawHeaders as unknown as HeadersInit | undefined,
          body,
        }),
      { url, method, rawHeaders, body }
    );
    // Response is not serialisable, so we must serialise it
    /* eslint-disable */
    const response = await responseHandle.evaluate((response, props) => {
      const responseJson: Record<string, any> = Object.fromEntries(
        props.map((key) => [key, response[key]])
      );
      responseJson.headers = Object.fromEntries(responseJson.headers.entries());
      return responseJson as SerialisedResponse;
    }, SERIALISABLE_RESPONSE_PROPS);
    /* eslint-enable */
    if (!response.ok) {
      throw new Error(`Response not OK: ${prettyPrint(response)}`);
    }
    this.logger.trace("Response: " + prettyPrint(response));
    return new FoodstuffsResponse(responseHandle, response);
  }

  private async isLoggedIn(page: Page) {
    type DataLayer = { loginState?: "loggedIn" | "guest" }[];
    const loginState = await page.evaluate(() => {
      const dataLayer = (globalThis as unknown as { dataLayer: DataLayer }).dataLayer;
      return dataLayer.find((entry) => "loginState" in entry)?.loginState;
    });
    return loginState === "loggedIn";
  }
}

export interface FoodstuffsUserProfile {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  agreedToMarketing: boolean;
  agreedToTermsAndConditions: boolean;
  clubCardUser: boolean;
}

export interface LoginResponse {
  success: boolean;
  userProfile: FoodstuffsUserProfile;
}

/**
 * Response backed by a Playwright JSHandle browser fetch Response. json()
 * and text() extraction methods are available.
 */
class FoodstuffsResponse implements Response {
  readonly bodyUsed: boolean;
  readonly headers: Headers;
  readonly ok: boolean;
  readonly redirected: boolean;
  readonly status: number;
  readonly statusText: string;
  readonly type: ResponseType;
  readonly url: string;

  constructor(
    private readonly responseHandle: JSHandle<Response>,
    responseJson: SerialisedResponse
  ) {
    this.bodyUsed = responseJson.bodyUsed;
    this.ok = responseJson.ok;
    this.redirected = responseJson.redirected;
    this.status = responseJson.status;
    this.statusText = responseJson.statusText;
    this.type = responseJson.type;
    this.url = responseJson.url;
    this.headers = new Headers(responseJson.headers);
  }

  /* eslint-disable */
  async json(): Promise<any> {
    const body = await this.text();
    try {
      return JSON.parse(body);
    } catch (error) {
      throw new Error("Failed to parse response body as JSON: \n" + body);
    }
  }
  /* eslint-enable */

  async text(): Promise<string> {
    return this.responseHandle.evaluate((response) => response.text());
  }

  get body(): ReadableStream<Uint8Array> | null {
    throw new Error("Cannot access body of FoodstuffsResponse directly");
  }
  get size(): number {
    throw new Error("Cannot access size of FoodstuffsResponse directly");
  }
  buffer(): Promise<Buffer> {
    throw new Error("Method not implemented.");
  }
  clone(): Response {
    throw new Error("Method not implemented.");
  }
  arrayBuffer(): Promise<ArrayBuffer> {
    throw new Error("Method not implemented.");
  }
  blob(): Promise<Blob> {
    throw new Error("Method not implemented.");
  }
  formData(): Promise<FormData> {
    throw new Error("Method not implemented.");
  }
}

const UNSERIALISABLE_RESPONSE_PROPS = [
  "arrayBuffer",
  "blob",
  "formData",
  "json",
  "text",
  "clone",
  "body",
] as const;
type UnserialisableResponseProp = typeof UNSERIALISABLE_RESPONSE_PROPS[number];

const SERIALISABLE_RESPONSE_PROPS = [
  "headers",
  "ok",
  "redirected",
  "status",
  "statusText",
  "type",
  "url",
] as const;

interface SerialisedResponse extends Omit<Response, UnserialisableResponseProp | "headers"> {
  headers: Record<string, string>;
}
