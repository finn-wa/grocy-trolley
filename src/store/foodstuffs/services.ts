import { getEnvAs } from "@gt/utils/environment";
import { Browser, firefox } from "playwright";
import { LogLevel, playwrightLogger } from "@gt/utils/logger";
import { FoodstuffsCartService } from "./cart/foodstuffs-cart-service";
import { FoodstuffsListService } from "./lists/foodstuffs-list-service";
import { FoodstuffsOrderService } from "./orders/foodstuffs-order-service";
import { FoodstuffsSearchService } from "./search/foodstuffs-search-service";
import { FoodstuffsUserAgent } from "./rest/foodstuffs-user-agent";

let browser: Browser | null = null;

/**
 * Returns a lazy-loaded shared browser instance.
 * @returns The browser instance
 */
export async function getBrowser(): Promise<Browser> {
  if (browser) {
    return browser;
  }
  browser = await firefox.launch({
    headless: true,
    logger: playwrightLogger(LogLevel.WARN),
  });
  return browser;
}

export async function foodstuffsServices(): Promise<FoodstuffsServices> {
  const userAgent = new FoodstuffsUserAgent(
    getBrowser,
    getEnvAs({ PAKNSAVE_EMAIL: "email", PAKNSAVE_PASSWORD: "password" })
  );
  return {
    userAgent,
    cartService: new FoodstuffsCartService(userAgent),
    listService: new FoodstuffsListService(userAgent),
    orderService: new FoodstuffsOrderService(userAgent),
    searchService: new FoodstuffsSearchService(userAgent),
  };
}

export interface FoodstuffsServices {
  userAgent: FoodstuffsUserAgent;
  cartService: FoodstuffsCartService;
  listService: FoodstuffsListService;
  orderService: FoodstuffsOrderService;
  searchService: FoodstuffsSearchService;
}
