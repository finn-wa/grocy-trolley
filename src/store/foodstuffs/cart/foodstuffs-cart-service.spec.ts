import { LoginDetails } from "@gt/store/shared/rest/login-details.model";
import { getEnvAs, initEnv } from "@gt/utils/environment";
import { FoodstuffsUserAgent } from "../rest/foodstuffs-user-agent";
import { getBrowser } from "../services";
import { FoodstuffsCartService } from "./foodstuffs-cart-service";
import { CartProductRef, FoodstuffsCart } from "./foodstuffs-cart.model";

describe("FoodstuffsCartService", () => {
  let cartService: FoodstuffsCartService;

  initEnv({
    envFilePath: ".test.env",
    envFilePathOptional: true,
    requiredVars: ["PAKNSAVE_EMAIL", "PAKNSAVE_PASSWORD"],
  });
  const loginDetails: LoginDetails = getEnvAs({
    PAKNSAVE_EMAIL: "email",
    PAKNSAVE_PASSWORD: "password",
  });
  const milk: CartProductRef = {
    productId: "5201479-EA-000",
    quantity: 1,
    sale_type: "UNITS",
  };
  const carrots: CartProductRef = {
    productId: "5039965-KGM-000",
    quantity: 500,
    sale_type: "WEIGHT",
  };

  beforeEach(async () => {
    const userAgent = new FoodstuffsUserAgent(getBrowser, loginDetails);
    cartService = new FoodstuffsCartService(userAgent);
    await cartService.clearCart();
  });

  describe("unit tests", () => {
    test("getCart", async () => {
      const cart = await cartService.getCart();
      expect(cart).toBeTruthy();
      expect(cart).toHaveProperty("products");
      expect(Array.isArray(cart.products)).toBe(true);
    });

    test("addProductsToCart", async () => {
      const emptyCart = await cartService.getCart();
      expect(emptyCart.products.length).toBe(0);

      const products = [milk, carrots];
      const fullCart = await cartService.addProductsToCart(products);

      expect(fullCart.products.length).toEqual(2);
      const productIdsOf = (products: CartProductRef[]) => products.map((p) => p.productId).sort();
      expect(productIdsOf(fullCart.products)).toEqual(productIdsOf(products));
    });

    test("clearCart", async () => {
      const fullCart = await cartService.addProductsToCart([milk, carrots]);
      expect(fullCart.products.length).toBe(2);

      const response = await cartService.clearCart();
      expect(response.success).toBe(true);
      const emptyCart = await cartService.getCart();
      expect(emptyCart.products.length).toBe(0);
    });
  });

  describe("snapshot tests", () => {
    /**
     * Overwrites changeable properties with fixed values for the snapshot.
     * @param cart Cart
     * @returns Snapshot-ready cart
     */
    function cartSnapshot(cart: FoodstuffsCart): FoodstuffsCart {
      return { ...cart, products: cart.products.map((product) => ({ ...product, price: 0 })) };
    }

    test("getCart", async () => {
      await cartService.addProductsToCart([milk, carrots]);
      const cart = await cartService.getCart();
      expect(cartSnapshot(cart)).toMatchSnapshot();
    });

    test("addProductsToCart", async () => {
      const cart = await cartService.addProductsToCart([milk, carrots]);
      expect(cartSnapshot(cart)).toMatchSnapshot();
    });

    test("clearCart", async () => {
      await cartService.addProductsToCart([milk, carrots]);
      expect(await cartService.clearCart()).toEqual({ success: true });
    });
  });
});
