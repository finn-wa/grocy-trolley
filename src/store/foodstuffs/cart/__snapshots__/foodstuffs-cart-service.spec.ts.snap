// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoodstuffsCartService snapshot tests addProductsToCart 1`] = `
Object {
  "allowSubstitutions": true,
  "bagFee": 100,
  "orderNumber": 0,
  "products": Array [
    Object {
      "badgeImageUrl": "/assets/images/PNS/Everyday_Low.svg",
      "catalogPrice": 0,
      "categoryName": "Fruit & Vegetables",
      "hasBadge": true,
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5039965.png",
      "liquor": false,
      "name": "Carrots",
      "originStatement": "Product of New Zealand",
      "price": 0,
      "productId": "5039965-KGM-000",
      "promoBadgeImageTitle": "Everyday Low",
      "promotionCode": "",
      "quantity": 500,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 200,
          "stepSize": 100,
          "type": "WEIGHT",
          "unit": "g",
        },
      ],
      "sale_type": "WEIGHT",
      "tobacco": false,
      "uom": "g",
      "weightDisplayName": "kg",
    },
    Object {
      "badgeImageUrl": "/assets/images/PNS/Everyday_Low.svg",
      "brand": "Value",
      "catalogPrice": 0,
      "categoryName": "Dairy & Eggs",
      "hasBadge": true,
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5201479.png",
      "liquor": false,
      "name": "Standard Milk",
      "price": 0,
      "productId": "5201479-EA-000",
      "promoBadgeImageTitle": "Everyday Low",
      "promotionCode": "",
      "quantity": 1,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 1,
          "stepSize": 1,
          "type": "UNITS",
          "unit": "ea",
        },
      ],
      "sale_type": "UNITS",
      "tobacco": false,
      "weightDisplayName": "2l",
    },
  ],
  "promoCodeDiscount": 0,
  "saving": 0,
  "serviceFee": 0,
  "store": Object {
    "storeAddress": "691 Manukau Road, Royal Oak, Auckland, 1023",
    "storeId": "e1925ea7-01bc-4358-ae7c-c6502da5ab12",
    "storeName": "PAK'nSAVE Royal Oak",
    "storeRegion": "NI",
  },
  "subtotal": 516,
  "unavailableProducts": Array [],
  "wasRepriced": false,
}
`;

exports[`FoodstuffsCartService snapshot tests getCart 1`] = `
Object {
  "allowSubstitutions": true,
  "bagFee": 100,
  "orderNumber": 0,
  "products": Array [
    Object {
      "badgeImageUrl": "/assets/images/PNS/Everyday_Low.svg",
      "catalogPrice": 0,
      "categoryName": "Fruit & Vegetables",
      "hasBadge": true,
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5039965.png",
      "liquor": false,
      "name": "Carrots",
      "originStatement": "Product of New Zealand",
      "price": 0,
      "productId": "5039965-KGM-000",
      "promoBadgeImageTitle": "Everyday Low",
      "promotionCode": "",
      "quantity": 500,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 200,
          "stepSize": 100,
          "type": "WEIGHT",
          "unit": "g",
        },
      ],
      "sale_type": "WEIGHT",
      "tobacco": false,
      "uom": "g",
      "weightDisplayName": "kg",
    },
    Object {
      "badgeImageUrl": "/assets/images/PNS/Everyday_Low.svg",
      "brand": "Value",
      "catalogPrice": 0,
      "categoryName": "Dairy & Eggs",
      "hasBadge": true,
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5201479.png",
      "liquor": false,
      "name": "Standard Milk",
      "price": 0,
      "productId": "5201479-EA-000",
      "promoBadgeImageTitle": "Everyday Low",
      "promotionCode": "",
      "quantity": 1,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 1,
          "stepSize": 1,
          "type": "UNITS",
          "unit": "ea",
        },
      ],
      "sale_type": "UNITS",
      "tobacco": false,
      "weightDisplayName": "2l",
    },
  ],
  "promoCodeDiscount": 0,
  "saving": 0,
  "serviceFee": 0,
  "store": Object {
    "storeAddress": "691 Manukau Road, Royal Oak, Auckland, 1023",
    "storeId": "e1925ea7-01bc-4358-ae7c-c6502da5ab12",
    "storeName": "PAK'nSAVE Royal Oak",
    "storeRegion": "NI",
  },
  "subtotal": 516,
  "unavailableProducts": Array [],
  "wasRepriced": false,
}
`;
